src/Pagination/Paginator.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace App\Pagination;
+
+use Doctrine\ORM\QueryBuilder as DoctrineQueryBuilder;
+use Doctrine\ORM\Tools\Pagination\CountWalker;
+use Doctrine\ORM\Tools\Pagination\Paginator as DoctrinePaginator;
+
+/**
+ * @author Javier Eguiluz <javier.eguiluz@gmail.com>
+ */
+class Paginator
+{
+    private const PAGE_SIZE = 10;
+    private $currentPage;
+    private $pageSize;
+    private $results;
+    private $numResults;
+
+    public function __construct(DoctrineQueryBuilder $queryBuilder, int $currentPage = 1, int $pageSize = self::PAGE_SIZE)
+    {
+        $this->currentPage = max(1, $currentPage);
+        $this->pageSize = $pageSize;
+        $firstResult = ($this->currentPage - 1) * $pageSize;
+
+        $query = $queryBuilder
+            ->setFirstResult($firstResult)
+            ->setMaxResults($pageSize)
+            ->getQuery();
+
+        if (0 === \count($queryBuilder->getDQLPart('join'))) {
+            $query->setHint(CountWalker::HINT_DISTINCT, false);
+        }
+
+        $paginator = new DoctrinePaginator($query, true);
+        $paginator->setUseOutputWalkers(false);
```

Comment:
This should detect whether the QueryBuilder has a `having` clause, as the tree walker cannot count those (the output worker is necessary in such case).
And it should also detect whether the root entity uses an association as identifier (in case we paginate with a fetch join condition), as it is not supported either by the tree walker.

Reply:
Thanks for this review and for the explanation. Sadly this is way above my knowledge of Doctrine ... so I'd need the exact needed changes in the code to add them. Thanks.

Reply:
Hm, IIRC we use this paginator in one spot only... if so - I don't think it should be configured unless it needed IMHO

Reply:
@javiereguiluz As we spoke extending doctrine's paginator is not an option. Maybe exposing it with a getter and leaving the developer to decide whether to turn this on/off is a better idea?

Reply:
even if this is a demo and there is no `having` clause currently in the code, I think we should make the adjustment according to Stof's comment, so anyone copying and pasting from here would have a correct version that covers this case.

Something like this could work:
```php
$paginator->setUseOutputWalkers(\count($this->queryBuilder->getDQLPart('having')) > 0);
```